design/ceph-openshift-fencing-mitigation.md

```diff
@@ -0,0 +1,72 @@
+
+## Goals
+
+- Use the MachineDisruptionBudget to ensure that OCP 4.x style fencing does not cause data unavailability and loss.
+
+## What is fencing in OCP 4.x?
+
+Openshift uses `Machines` and `MachineSets` from the [cluster-api](https://github.com/kubernetes-sigs/cluster-api) to dynamically provisions nodes. Fencing is a remediation method that reboots/deletes `Machine` CRDs to solve problems with automatically provisioned nodes.
+
+Once [MachineHealthCheck controller](https://github.com/openshift/machine-api-operator#machine-healthcheck-controller) detects that a node is `NotReady` (or some other configured condition), it will remove the associated `Machine` which will cause the node to be deleted. The `MachineSet` controller will then replace the `Machine` via the machine-api. The exception is on baremetal platforms where fencing will reboot the underlying `BareMetalHost` object instead of deleting the `Machine`.
+
+
+## Why can't we use `PodDisruptionBudget`?
+
+Fencing does not use the eviction api. It is for `Machine`s and not `Pod`s.
+
+## Will we need to do large storage rebalancing after fencing?
+
+Hopefully not. On cloud platforms, the OSDs can be rescheduled on new nodes along with their backing PVs, and on baremetal where the local PVs are tied to a node, fencing will simply reboot the node instead of destroying it.
+
+
+# Problem Statement
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has PGs clean) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated as MachineHealthCheck does not check anything other than the node state.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as well as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy > currentHealthy, we can disable fencing on the nodes the MDB points to.
+
+# Design:
+
+We will implement two controllers `machinedisruptionbudget-controller` and the `machine-controller` to be implemented through the controller pattern describere [here](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg#hdr-Controller_Writing_Tips). Each controller watches a set of object kinds and reconciles one.
+
+The bottom line is that fencing is blocked if the PG state in not active+clean, but fencing continues on `Machine`s without the label which indicates that OSD resources are running there.
+
+## machinedisruptionbudget-controller
+This controller watches ceph PGs and CephClusters. We will ensure the reconciler is enqueued every 60s. It ensures that each CephCluster has a MDB created, and the MDB's value of maxUnvailable reflects the health of the Ceph Cluster's PGs.
+If all PGs are clean, maxUnavailable = 1.
+else, maxUnavailable = 0.
```

Comment:
You can do it this way, but the hope was that on metal, "bad" nodes would loose the label associated with the MDB.  This would result in minAvailable being higher than the current number of healthy nodes and prevent fencing.  Then, once the OSD was both running and up-to-date, it would get the label back again and fencing would again be possible.

Reply:
@beekhof We do remove the label from bad nodes. See `machine-controller` under the same document. We just use this mechanism to disable the ordinarily allowed fencing of good(wrt ceph) nodes when ceph is not resynced. It is a common scenario after initial fencing for all the nodes to be good, but ceph is not ready to lose any of them until they sync.

Reply:
Or do you mean it is better to always have maxUnavailable = 0 and only allow fencing by removing labels? That would simplify things, but we need to be able to assume that whatever problem makes fencing necessary on the node also impacts Ceph; otherwise, fencing would be blocked indefinitely.